src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java

```diff
@@ -96,8 +97,18 @@ public static SearchRequest parseSearchRequest(RestRequest request) {
             }
         }
 
+        // do not allow 'query_and_fetch' or 'dfs_query_and_fetch' search types
+        // from the REST layer. these modes are an internal optimization and should
+        // not be specified explicitly by the user.
+        String searchType = request.param("search_type");
+        if (SearchType.fromString(searchType).equals(SearchType.QUERY_AND_FETCH) ||
+                SearchType.fromString(searchType).equals(SearchType.DFS_QUERY_AND_FETCH)) {
+            throw new ElasticsearchIllegalArgumentException("Unsupported search type [" + searchType + "]");
```

Comment:
I'm a bit nitpicking but if we can avoid calling this parsing method twice easily I think we should do it? Additionally I tend to find that extracting intermediate results to local variables helps readability.